### DISCUSSION TITLE
Volume setting (nearly) always at 100%

### DISCUSSION THREAD

**Discussion by firstchanceongit** [rern/rAudio#739] [Fixed Issues]
Having used a pre-1 version of rAudio for a little while I made a fresh installation of rAudio today (incl. most recent update). I have not yet tested all of it but one thing I'm severly struggling with is the volume setting which always is at 100%. (I've seen there were a couple of posts upon this (or at least a similar) issue from about a year ago but it's not clear to me from them whether the issue got solved and how.)

My setup:
rAudio 1 • 20220107
5.10.82-4 • armv7l
Raspberry Pi 2B Rev 1.1
DAC: HifiBerry AMP+

Under "Settings/GPIO Devices" I selected "HifBerry Amp" (since no "HifiBerry AMP+" available but not sure there is a difference; AMP2 is definitely a different DAC).

Under "Player/Output" my selections are:
Device: HifiBerry AMP
Mixer Device: Master
Volume Control: Mixer device
The setting behind the speaker symbol is 100 and jumps back to this in case I try to lower it.
Neither the switch Equalizer nor any of the switches under Bit-Perfect, Volume or Options are selected.

When playing a music file, the volume knob stands at 100 (and jumps back to this if I try to lower it).

Sometimes I can level the volume down with the Minus button but the volume remains to stand at 100 which is rather confusing.

What am I missing here? It all worked fine and logical in earlier versions and was fully controllable via the knob, always showing and keeping the current volume setting.

**Comment by rern** [comment]
Let's see: Player > click Output - copy data from the box and post.

**Comment by firstchanceongit** [comment]
Currently, I cannot enter any of the right-hand menus at all (raudio symbol never stops pulsing). I need to check again this evening.

**Comment by firstchanceongit** [comment]
I had to make a fresh installation (incl. most recent update, incl. from update branch) since I could not switch into the Player menu. Same behaviour as before, settings as stated above. Player output:

# cat /etc/asound.conf
defaults.pcm.card 0
defaults.ctl.card 0

# aplay -l | grep ^card
card 0: sndrpihifiberry [snd_rpi_hifiberry_amp], device 0: HifiBerry AMP HiFi tas5713.1-001b-0 [HifiBerry AMP HiFi tas5713.1-001b-0]

# amixer scontrols


And just in case it is useful, here also the content of MPD:

mpd 0.23.5-1
bind_to_address        "0.0.0.0"
db_file                "/srv/http/data/mpd/mpd.db"
music_directory        "/mnt/MPD"
playlist_directory     "/var/lib/mpd/playlists"
replaygain             "off"
state_file             "/var/lib/mpd/mpdstate"
user                   "mpd"

input {
	plugin         "curl"
}

decoder {
	plugin         "ffmpeg"
	enabled        "no"
}

resampler {
	plugin         "soxr"
	quality        "very high"
}

audio_output {
	name           "HiFiBerry Amp"
	device         "hw:0,0"
	type           "alsa"
	auto_resample  "no"
	auto_format    "no"
	mixer_type     "hardware"
	mixer_control  "Master"
	mixer_device   "hw:0"
}

● mpd.service - Music Player Daemon
     Loaded: loaded (/usr/lib/systemd/system/mpd.service; disabled; vendor preset: disabled)
    Drop-In: /etc/systemd/system/mpd.service.d
             └─override.conf
     Active: active (running) since Mon 2022-01-10 19:49:19 CET; 2h 6min ago
       Docs: man:mpd(1)
             man:mpd.conf(5)
   Main PID: 3015 (mpd)
      Tasks: 3 (limit: 2129)
        CPU: 1.432s
     CGroup: /system.slice/mpd.service
             └─3015 /usr/bin/mpd --no-daemon

Jan 10 19:49:18 rAudio systemd[1]: Starting Music Player Daemon...
Jan 10 19:49:19 rAudio taskset[3015]: decoder: Decoder plugin 'wildmidi' is unavailable: configuration file does not exist: /etc/timidity/timidity.cfg
Jan 10 19:49:19 rAudio systemd[1]: Started Music Player Daemon.

**Reply by rern** [reply]
- Try: `alsamixer`
- Let me see the output of: `amixer scontrols`

**Comment by firstchanceongit** [comment]
amixer scontrols reports:
Simple mixer control 'Master',0
Simple mixer control 'Channels',0

**Comment by rern** [comment]
With latest update, is it OK now?

**Comment by firstchanceongit** [comment]
Unfortunately, no. When I change the volume, the knob miraculously slides back to 100%. Same if I try under Player -> Output -> Setting behind the speaker symbol.

**Reply by rern** [reply]
Let me see output of: `amixer`

**Comment by firstchanceongit** [comment]
Simple mixer control 'Master',0
  Capabilities: volume volume-joined
  Playback channels: Mono
  Capture channels: Mono
  Limits: 0 - 248
  Mono: 213 [86%]
Simple mixer control 'Channels',0
  Capabilities: volume
  Playback channels: Front Left - Front Right
  Capture channels: Front Left - Front Right
  Limits: 0 - 248
  Front Left: 201 [81%]
  Front Right: 201 [81%]

**Reply by rern** [reply]
Try:
- Without playing
- SSH terminal > `alsamixer`
- `Arrow down` key to 20
- `F6` > select Channels
- `Arrow down` key to 20
-
- Browser > play
- 
- SSH terminal
- `Arrow up` /  `Arrow down` key
- `F6` > select Master
- `Arrow up` /  `Arrow down` key

**Comment by firstchanceongit** [comment]
Nothing happens. And I hear no sond at all anylonger, even not too loud. Just nothing (browser still shows volume at 100).

**Reply by rern** [reply]
Let me see your `alsamixer` screen.
`alsamixer` should be able to control the volume.

**Comment by firstchanceongit** [comment]
![grafik](https://user-images.githubusercontent.com/78300209/151657695-0bc7e862-b49f-424d-b340-72e199fef5cc.png)

**Reply by rern** [reply]
While a track is playing, on `alsamixer` screen try:
- With arrow up key, increase volume until you hear the playing track.
- If no sound with Master, use arrow right to switch to Channels and try the same.
- If still no, lower both to 50 then press M to toggle mute/unmute.

If `alsamixer` could not control the volume:
- If your card is a generic or unnamed, you have to search for which specific overlay/driver to use.
- Or set Volume Control to software instead.

**Comment by firstchanceongit** [comment]
Tried. I'm getting sound when I move both, Master and Channels, into the area above, say, 50. With 70 on both I get reasonable volume for a normal room. (However, there seems to bee no connection to the browser volume knob et al. which always stays at 100 as stated in the postings above.)

**Reply by rern** [reply]
All right. One step closer.
Now get a revised file:
```sh
wget https://github.com/rern/rAudio-1/raw/UPDATE/srv/http/bash/cmd.sh -O /srv/http/bash/cmd.sh
```
Then try changing volume with the browser.

**Comment by firstchanceongit** [comment]
Yeah! Works like a charm :-)

Thanks a lot, rern!

**Comment by firstchanceongit** [comment]
Just one further observation (not sure whether connected to the above or not):

It seems not be possible to adjust the volume via browser per the +/- buttons in increments of 1. When hitting + or -, volume changes in larger steps of about 3. I do not not recollect this behaviour from earlier versions of raudio/runeaudio.

**Comment by rern** [comment]
It's the conversion from `%` to `dB`, the actual unit of volume control in logarithmic scale, which depends on `dB` min, max and resolution defined by the card's driver. So `+` / `-` = 1 step which may not be 1% if there're less than 100 `dB` steps.
Software volume, on the other hand, can always be at 1% precision.
rAudio uses `amixer` to control hardware volume directly, skip MPD volume control. With tap or drag, rAudio will get actual level once set and adjust displayed value accordingly.

**Comment by firstchanceongit** [comment]
Ok, understood. Thank you for the explanation (as well). Then that seems also to be the reason why the volume slider somtimes "magically" moves one step up or down after being set at a certain level.
